clang/docs/checkedc/rfc/Include-Checked-Hdrs.md

```diff
@@ -0,0 +1,182 @@
+
+             #ifdef IMPLICIT_INCLUDE_CHECKED_HDRS
+             #include <foo_checked.h>
+             #else
+             #include_next <foo.h>
+             #endif
+     - In `foo_checked.h`, we modify `#include <foo.h>` to
+       `#include_next <foo.h>`.
+
+     - The way this solution will work:
+        - If a program includes `foo.h`, clang will pick up either
+          `foo_checked.h` or the system `foo.h` depending on whether or not
+          `-DIMPLICIT_INCLUDE_CHECKED_HDRS` is specified on the compilation
+          commandline.  Therefore specifying this flag will cause the implicit
+          inclusion of the checked counterpart of `foo.h`, which will make it
+          convenient for automation. Not specifying the commandline flag will
+          not perform implicit inclusion, providing a mechanism to opt out of 
+          implicit inclusion. Infinite recursion is avoided because of
+	  `#include_next`.
+        - If a program includes `foo_checked.h` current behavior will prevail.
+          The above flag has no effect on the explicit inclusion of checked
+          header files.
+
+     - **Pros**: 1) The solution provides a way to opt into implicit inclusion
+       of checked header files. 2) The existing approach of explicit inclusion
+       is supported. 3) In most cases integration with the build system is easy:
+       the flag `-DIMPLICIT_INCLUDE_CHECKED_HDRS` is passed to the compilation
+       of all source files through the `CFLAGS` variable. 4) No changes are
+       required to the order in which include directories are searched. 5) It
+       is convenient for automation.
+
+     - **Cons**: 1) In some cases integration with the build system will be more
+       involved if the above flag needs to be passed to the compilation of most
+       source files and avoided for a few source files that may want to directly
+       include system header files in order to avoid Checked-C-specific
+       declarations. 2) It is difficult to accomodate clang-specific
+       declarations.
+
+## Proposed Solution:
+
+The proposed solution is as follows:
+  - We will implicitly include checked header files by default and provide the 
+    compilation flag NO_IMPLICIT_INCLUDE_CHECKED_HDRS to opt out of the implicit
+    inclusion.
+  - For header files that do not have clang-specific declarations, but have
+    Checked-C-specific declarations (there are 13 such header files at present),
+    we will do the following:
+      - For each system header file, say `foo.h`, that does not have
+        clang-specific declarations but has Checked-C-specific declarations, we
+        will add a new file also called `foo.h` that will contain the following:
+
+            #if !defined __checkedc || defined NO_IMPLICIT_INCLUDE_CHECKED_HDRS
+            #include_next <foo.h>
+            #else
+            #include <foo_checked.h>
+            #endif
+
+      - The file `foo_checked.h` will contain `#include_next <foo.h>` and the
```

Comment:
You could change the `#include_next <foo.h>` to `#include <foo.h>` (making it consistent with the case where Clang-specific declarations are present) if you add a guard to `foo.h` to avoid direct recursion between the two files.  (Otherwise, `foo.h` does not _need_ a guard because all it does is include two other files that presumably have their own guards.  But if it's considered best practice for `foo.h` to have a guard anyway, then you could adopt this suggestion.)